sources/cfsignal/src/main.rs

```diff
@@ -0,0 +1,97 @@
+#![deny(rust_2018_idioms, unused_imports)]
+
+/*!
+## Introduction
+
+Cfsignal is similar to [cfn-signal] (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-signal.html).
+
+When creating an Auto Scaling Group, CloudFormation can be configured to wait for the expected number of signals from instances before considering the ASG successfully created. See [CreationPolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-creationpolicy.html) and [UpdatePolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html) for more details.
+
+Cfsignal uses `systemctl is-system-running` to determine whether the boot has succeeded or failed, and sends the corresponding signal to the CloudFormation stack.
+
+## configuration
+
+Cfsignal is configured using the following settings:
+* `settings.cloudformation.should-signal`: Check status and send signal. Defaults to `false`.
+* `settings.cloudformation.stack-name`: Name of the stack to signal.
+* `settings.cloudformation.logical-resource-id`: The logical ID of the AutoScalingGroup resource that you want to signal.
+*/
+
+#![deny(rust_2018_idioms)]
+
+mod cloudformation;
+mod config;
+mod error;
+mod system_check;
+
+use crate::config::Config;
+use crate::error::Result;
+use crate::system_check::SystemCheck;
+use cloudformation::signal_resource;
+use log::LevelFilter;
+use log::{error, warn, info};
+use simplelog::{Config as LogConfig, SimpleLogger};
+use std::fs;
+use std::process;
+
+// We only want to run early-boot-config once, at first boot.  Our systemd unit file has a
+// ConditionPathExists that will prevent it from running again if this file exists.
+// We create it after running successfully.
+const MARKER_FILE: &str = "/var/lib/bottlerocket/cfsignal.ran";
+
+
+/// pub(crate) for testing.
+async fn run() -> Result<()> {
+    SimpleLogger::init(LevelFilter::Info, LogConfig::default())
+        .expect("unable to configure logger");
+
+    // load the cfsignal config file
+    let config = Config::from_file("/etc/cfsignal.toml")?;
+
+    let mut signal_status = "FAILURE";
+    let system_check = Box::new(SystemCheck {});
+    let system_status = system_check.system_running()?;
+    info!(
+        "System status is: {} [{}]",
+        system_status.status, system_status.exit_code
+    );
+
+    // run only if the opt-in flag is set
+    if config.should_signal {
+
```

Comment:
Since we only do anything with the system check status if `should_signal` is set, we can have the check inside the if block.
```suggestion


    // run only if the opt-in flag is set
    if config.should_signal {
        let system_check = Box::new(SystemCheck {});
        let system_status = system_check.system_running()?;
        info!(
            "System status is: {} [{}]",
            system_status.status, system_status.exit_code
        );
```

Reply:
Can be useful to know which would be the system status even without signaling it to CloudFormation, maybe to try it out in a `dry mode` or to debug.